modules/types/include/hephaestus/types/bounds.h

```diff
@@ -0,0 +1,100 @@
+//=================================================================================================
+// Copyright (C) 2023-2024 HEPHAESTUS Contributors
+//=================================================================================================
+
+#pragma once
+
+#include <concepts>
+#include <cstdint>
+#include <iostream>
+#include <random>
+
+#include <fmt/ostream.h>
+
+#include "hephaestus/random/random_object_creator.h"
+#include "hephaestus/utils/concepts.h"
+#include "hephaestus/utils/exception.h"
+
+namespace heph::types {
+
+enum class BoundsType : uint8_t {
+  INCLUSIVE,   // []
+  LEFT_OPEN,   // (]
+  RIGHT_OPEN,  // [)
+  OPEN         // ()
+};
+
+/// @brief A class representing range bounds.
+template <NumericType T>
+struct Bounds {
+  [[nodiscard]] inline constexpr auto operator==(const Bounds&) const -> bool = default;
+
+  [[nodiscard]] static auto random(std::mt19937_64& mt) -> Bounds;
+
+  [[nodiscard]] inline constexpr auto isWithinBounds(T value) const -> bool;
+
+  T lower_bound{};
+  T upper_bound{};
+  BoundsType bounds_type{ BoundsType::INCLUSIVE };
+};
+
+template <NumericType T>
+auto operator<<(std::ostream& os, const Bounds<T>& bounds) -> std::ostream&;
+
+//=================================================================================================
+// IMPLEMENTATION
+//=================================================================================================
+
+template <NumericType T>
+auto Bounds<T>::random(std::mt19937_64& mt) -> Bounds {
+  return { .lower_bound = random::random<T>(mt),
+           .upper_bound = random::random<T>(mt),
+           .bounds_type = random::random<BoundsType>(mt) };
+}
+
+template <NumericType T>
+inline constexpr auto Bounds<T>::isWithinBounds(T value) const -> bool {
```

Comment:
nit: wouldn't it be nicer to have this as a free function?

Reply:
Let's see what API we prefer. Currently, we have
```
template <typename T>
class Bounds {
  auto isWithingBounds(T value) -> bool;
};

void main() {
  Bouds<float> current_bounds_amps = { 0.0f, 0.5f };
  float current_amps = 0.3f;
  if(!isWithinBounds(current_amps, current_bounds_amps)) {
    // do something
  }
}
```
your suggestion would mean
```
template <typename T>
class Bounds {
};

template <typename T>
auto isWithinBounds(T value, const Bounds<T>& bounds) -> bool;

Bouds<float> current_bounds_amps = { 0.0f, 0.5f };
float current_amps = 0.3f;
if(!current_bounds_amps.isWithinBounds(current_amps)) {
  // do something
}
```
Give me a vote please @floriantschopp @filippobrizzi

Reply:
I kind of like @floriantschopp suggestion tbh, it reads more natural.

Reply:
let's go free function

Reply:
Free function it is!